File: pyref/atlasopt/decomposition.py
"""木分解(min-fill ヒューリスティック)と妥当性検証。フェーズ3 ステージ1。

分解はグラフのノード・辺集合のみに依存するため、データ(リーグ)更新時に
一度計算すればよい。terminal 指定ごとの再計算は不要。

min-fill: 「消去したとき近傍間に張る必要がある補充辺が最少の頂点」を貪欲に消す。
タイは (fill, degree, id) で解消する(決定的。TS 移植と一致させるため)。
bag の木は「消去頂点の近傍のうち最も早く消える頂点の bag を親にする」標準構成で、
最後に root ノードを含む bag へ根を付け替える(DP の終端条件のため)。
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class TreeDecomposition:
    bags: tuple[frozenset[str], ...]
    parent: tuple[int, ...]  # 各 bag の親 bag index(根は -1)
    root_bag: int  # root ノードを含む bag(分解木の根)
    width: int  # max |bag| - 1


def build_decomposition(adj: dict[str, frozenset[str]], root: str) -> TreeDecomposition:
    if root not in adj:
        raise ValueError(f"root {root} not in graph")
    nbrs: dict[str, set[str]] = {v: set(ns) for v, ns in adj.items()}

    def fill_count(v: str) -> int:
        ns = nbrs[v]
        cnt = 0
        for a in ns:
            cnt += len(ns - nbrs[a]) - 1  # a 自身は常に非隣接扱いになるため引く
        return cnt // 2

    fills = {v: fill_count(v) for v in nbrs}
    order: list[str] = []
    bags: list[frozenset[str]] = []
    bag_nbrs: list[set[str]] = []  # 消去時点の近傍(親決定用)
    remaining = set(nbrs)
    while remaining:
        v = min(remaining, key=lambda x: (fills[x], len(nbrs[x]), x))
        ns = set(nbrs[v])
        order.append(v)
        bags.append(frozenset({v} | ns))
        bag_nbrs.append(ns)
        for a in ns:
            nbrs[a] |= ns - {a}
            nbrs[a].discard(v)
        del nbrs[v]
        remaining.discard(v)
        # fill 値が変わりうるのは近傍とその近傍だけ
        dirty = set(ns)
        for a in ns:
            dirty |= nbrs[a]
        for a in dirty & remaining:
            fills[a] = fill_count(a)

    # 消去順による木(forest)を作り、root ノードを含む bag へ根を付け替える
    pos = {v: i for i, v in enumerate(order)}
    tree_adj: list[set[int]] = [set() for _ in bags]
    for i, ns in enumerate(bag_nbrs):
        if ns:
            p = pos[min(ns, key=lambda a: pos[a])]
            tree_adj[i].add(p)
            tree_adj[p].add(i)

    root_bag = min(i for i, b in enumerate(bags) if root in b)
    parent = [-1] * len(bags)
    seen = {root_bag}
    stack = [root_bag]
    while stack:
        i = stack.pop()
        for j in tree_adj[i]:
            if j not in seen:
                seen.add(j)
                parent[j] = i
                stack.append(j)
    if len(seen) != len(bags):
        raise AssertionError("decomposition tree is not connected")  # 連結グラフでは起きない

    width = max(len(b) for b in bags) - 1
    return TreeDecomposition(
        bags=tuple(bags), parent=tuple(parent), root_bag=root_bag, width=width
    )


def verify(
    adj: dict[str, frozenset[str]], td: TreeDecomposition, root: str | None = None
) -> list[str]:
    """木分解の3条件を検証し、問題点のリストを返す(空なら妥当)。"""
    problems: list[str] = []
    n_bags = len(td.bags)

    # 親配列が根 root_bag の木であること
    if not 0 <= td.root_bag < n_bags:
        problems.append("root_bag out of range")
    elif td.parent[td.root_bag] != -1:
        problems.append("root_bag has a parent")
    if sum(1 for p in td.parent if p == -1) != 1:
        problems.append("parent array has multiple roots")
    if root is not None and root not in td.bags[td.root_bag]:
        problems.append("root node not in root_bag")

    # (1) 全頂点被覆
    covered: set[str] = set()
    for b in td.bags:
        covered |= b
    missing = set(adj) - covered
    if missing:
        problems.append(f"vertices not covered: {sorted(missing)[:5]}...")

    # (2) 全辺被覆
    for u, ns in adj.items():
        for v in ns:
            if u < v and not any(u in b and v in b for b in td.bags):
                problems.append(f"edge not covered: {u}-{v}")

    # (3) running intersection: 各頂点を含む bag 集合が分解木上で連結
    #     連結 ⇔ (bag数) - (親子とも当該頂点を含む木辺数) == 1
    bags_of: dict[str, list[int]] = {}
    for i, b in enumerate(td.bags):
        for v in b:
            bags_of.setdefault(v, []).append(i)
    for v, idxs in bags_of.items():
        edges = sum(1 for i in idxs if td.parent[i] != -1 and v in td.bags[td.parent[i]])
        if len(idxs) - edges != 1:
            problems.append(f"running intersection violated for {v}")

    if td.width != max(len(b) for b in td.bags) - 1:
        problems.append("width mismatch")
    return problems

File: web/src/solver/decomposition.ts
/**
 * 木分解(min-fill ヒューリスティック)と妥当性検証。
 * pyref/atlasopt/decomposition.py の移植(あちらが仕様原本)。
 *
 * 消去順・親の決め方・根の付け替えまで pyref と決定的に一致させている
 * (タイは (fill, degree, id) の辞書式。id 比較は ASCII なので Python と同順)。
 * 分解はグラフ構造のみに依存するため worker 初期化時に一度計算すればよい。
 */

export interface TreeDecomposition {
  /** 各 bag の頂点(昇順) */
  readonly bags: readonly (readonly string[])[];
  /** 各 bag の親 index(根は -1) */
  readonly parent: readonly number[];
  /** root ノードを含む bag(分解木の根) */
  readonly rootBag: number;
  readonly width: number;
}

type Adj = ReadonlyMap<string, ReadonlySet<string>>;

export function buildDecomposition(adj: Adj, root: string): TreeDecomposition {
  if (!adj.has(root)) throw new Error(`root ${root} not in graph`);
  const nbrs = new Map<string, Set<string>>();
  for (const [v, ns] of adj) nbrs.set(v, new Set(ns));

  const fillCount = (v: string): number => {
    const ns = nbrs.get(v)!;
    let cnt = 0;
    for (const a of ns) {
      const an = nbrs.get(a)!;
      let missing = 0;
      for (const b of ns) if (b !== a && !an.has(b)) missing++;
      cnt += missing;
    }
    return cnt / 2;
  };

  const fills = new Map<string, number>();
  for (const v of nbrs.keys()) fills.set(v, fillCount(v));

  const order: string[] = [];
  const bags: string[][] = [];
  const bagNbrs: Set<string>[] = [];
  const remaining = new Set(nbrs.keys());
  while (remaining.size > 0) {
    // (fill, degree, id) の辞書式最小(id で全順序なので一意)
    let v: string | null = null;
    for (const x of remaining) {
      if (v === null) {
        v = x;
        continue;
      }
      const fx = fills.get(x)!;
      const fv = fills.get(v)!;
      const dx = nbrs.get(x)!.size;
      const dv = nbrs.get(v)!.size;
      if (fx < fv || (fx === fv && (dx < dv || (dx === dv && x < v)))) v = x;
    }
    const ns = new Set(nbrs.get(v!)!);
    order.push(v!);
    bags.push([v!, ...ns].sort());
    bagNbrs.push(ns);
    for (const a of ns) {
      const an = nbrs.get(a)!;
      for (const b of ns) if (b !== a) an.add(b);
      an.delete(v!);
    }
    nbrs.delete(v!);
    remaining.delete(v!);
    // fill 値が変わりうるのは近傍とその近傍だけ
    const dirty = new Set(ns);
    for (const a of ns) for (const b of nbrs.get(a)!) dirty.add(b);
    for (const a of dirty) if (remaining.has(a)) fills.set(a, fillCount(a));
  }

  // 消去順による木(forest)を作り、root ノードを含む bag へ根を付け替える
  const pos = new Map(order.map((v, i) => [v, i] as const));
  const treeAdj: Set<number>[] = bags.map(() => new Set());
  bagNbrs.forEach((ns, i) => {
    if (ns.size > 0) {
      let p = -1;
      for (const a of ns) {
        const pa = pos.get(a)!;
        if (p === -1 || pa < p) p = pa;
      }
      treeAdj[i]!.add(p);
      treeAdj[p]!.add(i);
    }
  });

  let rootBag = -1;
  for (let i = 0; i < bags.length; i++) {
    if (bags[i]!.includes(root)) {
      rootBag = i;
      break;
    }
  }
  const parent = new Array<number>(bags.length).fill(-1);
  const seen = new Set([rootBag]);
  const stack = [rootBag];
  while (stack.length > 0) {
    const i = stack.pop()!;
    for (const j of treeAdj[i]!) {
      if (!seen.has(j)) {
        seen.add(j);
        parent[j] = i;
        stack.push(j);
      }
    }
  }
  if (seen.size !== bags.length) {
    throw new Error("decomposition tree is not connected"); // 連結グラフでは起きない
  }

  const width = Math.max(...bags.map((b) => b.length)) - 1;
  return { bags, parent, rootBag, width };
}

/** 木分解の3条件を検証し、問題点のリストを返す(空なら妥当)。 */
export function verify(adj: Adj, td: TreeDecomposition, root?: string): string[] {
  const problems: string[] = [];
  const nBags = td.bags.length;
  const bagSets = td.bags.map((b) => new Set(b));

  if (!(td.rootBag >= 0 && td.rootBag < nBags)) problems.push("rootBag out of range");
  else if (td.parent[td.rootBag] !== -1) problems.push("rootBag has a parent");
  if (td.parent.filter((p) => p === -1).length !== 1) {
    problems.push("parent array has multiple roots");
  }
  if (root !== undefined && !bagSets[td.rootBag]!.has(root)) {
    problems.push("root node not in rootBag");
  }

  // (1) 全頂点被覆
  const covered = new Set<string>();
  for (const b of td.bags) for (const v of b) covered.add(v);
  for (const v of adj.keys()) {
    if (!covered.has(v)) problems.push(`vertex not covered: ${v}`);
  }

  // (2) 全辺被覆
  for (const [u, ns] of adj) {
    for (const v of ns) {
      if (u < v && !bagSets.some((b) => b.has(u) && b.has(v))) {
        problems.push(`edge not covered: ${u}-${v}`);
      }
    }
  }

  // (3) running intersection: 連結 ⇔ (bag数) - (親子とも含む木辺数) == 1
  const bagsOf = new Map<string, number[]>();
  bagSets.forEach((b, i) => {
    for (const v of b) {
      if (!bagsOf.has(v)) bagsOf.set(v, []);
      bagsOf.get(v)!.push(i);
    }
  });
  for (const [v, idxs] of bagsOf) {
    const edges = idxs.filter(
      (i) => td.parent[i]! !== -1 && bagSets[td.parent[i]!]!.has(v),
    ).length;
    if (idxs.length - edges !== 1) problems.push(`running intersection violated for ${v}`);
  }

  if (td.width !== Math.max(...td.bags.map((b) => b.length)) - 1) {
    problems.push("width mismatch");
  }
  return problems;
}
